### PR TITLE
7830 multi schema support

### PR DESCRIPTION
**Pull request by nineinchnick**
Resolves yiisoft/yii2#7830. Adds support for relations between tables in different schemas.

### CODE REVIEW

File: generators/model/Generator.php [hunk 1]
```diff
@@ -287,6 +290,45 @@ public function generateRules($table)
     }
 
     /**
+     * @param \yii\db\TableSchema the table being checked
```
Comment (samdark) [inline generators/model/Generator.php]: Description of what method does is required,

File: generators/model/Generator.php [hunk 2]
```diff
@@ -287,6 +290,45 @@ public function generateRules($table)
...
+     * @param array $fks
+     * @param array $relations
+     * @return array
+     */
+    private function generateManyManyRelations($table, $fks, $relations)
+    {
+        $db         = $this->getDbConnection();
```
Comment (samdark) [inline generators/model/Generator.php]: Generally we aren't aligning assignments like it's done here.

File: generators/model/Generator.php [hunk 3]
```diff
@@ -287,6 +290,45 @@ public function generateRules($table)
...
+        $table0     = $fks[$table->primaryKey[0]][0];
```
Comment (samdark) [inline generators/model/Generator.php]: Could it be named better than `$table0`? Perharps `$pkPrimaryTable`?

File: generators/model/Generator.php [hunk 4]
```diff
@@ -297,79 +339,65 @@ protected function generateRelations()
...
-                        $hasMany = false;
-                        break;
+        foreach ($schemaNames as $schemaName) {
+            foreach ($db->getSchema()->getTableSchemas($schemaName) as $table) {
+                $className = $this->generateClassName($table->fullName);
+                foreach ($table->foreignKeys as $refs) {
+                    $refTable = $refs[0];
+                    $refTableSchema = $db->getTableSchema($refTable);
+                    unset($refs[0]);
+                    $fks          = array_keys($refs);
```
Comment (samdark) [inline generators/model/Generator.php]: Better not to align assignments. It's no consistent with overall code style.

File: generators/model/Generator.php [hunk 5]
```diff
@@ -297,79 +339,65 @@ protected function generateRelations()
...
-
-            $link = $this->generateRelationLink([$fks[$table->primaryKey[1]][1] => $table->primaryKey[1]]);
-            $viaLink = $this->generateRelationLink([$table->primaryKey[0] => $fks[$table->primaryKey[0]][1]]);
-            $relationName = $this->generateRelationName($relations, $className0, $db->getTableSchema($table0), $table->primaryKey[1], true);
-            $relations[$className0][$relationName] = [
-                "return \$this->hasMany($className1::className(), $link)->viaTable('" . $this->generateTableName($table->name) . "', $viaLink);",
-                $className1,
-                true,
-            ];
+                if (($fks = $this->checkPivotTable($table)) === false) {
```
Comment (samdark) [inline generators/model/Generator.php]: As I remember, correct name for the "middle" table is "junction table", not "pivot table". The latter is something else.

File: generators/model/Generator.php (hunk 3 above)
Comment (nineinchnick) [inline generators/model/Generator.php]: This is the old code, I've just moved it from `generateRelations()` to a separate method. I've got another issue created about this so I'm going to refactor it later.

File: generators/model/Generator.php (hunk 5 above)
Comment (nineinchnick) [inline generators/model/Generator.php]: Again, I just changed the indentation here, the naming remains. I also like "junction table" better but I don't want to introduce unecessary changes.

File: generators/model/Generator.php (hunk 3 above)
Comment (samdark) [inline generators/model/Generator.php]: OK.

File: generators/model/Generator.php (hunk 5 above)
Comment (samdark) [inline generators/model/Generator.php]: OK.

**Comment by nineinchnick** [comment]
Oh, forgot to mention that this PR requires yiisoft/yii#8012 to be accepted first, that's why builds fail now.

**Comment by samdark** [comment]
8012 merged.

**Comment by samdark** [comment]
I think it's not the reason for build failure: https://travis-ci.org/yiisoft/yii2-gii/jobs/57717304#L205. See the message that SQLite doesn't have anything about schema...

**Comment by nineinchnick** [comment]
Tests run against yii that's in composer.json and it's 2.0.3. Also yiisoft/yii2#7973 from upcoming 2.0.4 is required here.

**Comment by samdark** [comment]
Right.

**Comment by samdark** [comment]
Yep, verified locally that tests are OK.

**Comment by samdark** [comment]
Thank you!

**Pull request merged**